xdsl/dialects/irdl.py

```diff
@@ -0,0 +1,389 @@
+"""Definition of the IRDL dialect."""
+
+from __future__ import annotations
+from typing import Annotated, Iterable, Sequence
+from xdsl.dialects.builtin import StringAttr, SymbolRefAttr
+from xdsl.ir import (
+    Attribute,
+    Block,
+    Dialect,
+    OpResult,
+    ParametrizedAttribute,
+    Region,
+    SSAValue,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    IRDLOperation,
+    OpAttr,
+    SingleBlockRegion,
+    VarOperand,
+    irdl_attr_definition,
+    irdl_op_definition,
+)
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.traits import HasParent
+
+################################################################################
+# Dialect, Operation, and Attribute definitions                                #
+################################################################################
+
+
+@irdl_attr_definition
+class AttributeType(ParametrizedAttribute, TypeAttribute):
+    """Type of a attribute handle."""
+
+    name = "irdl.attribute"
+
+
+@irdl_op_definition
+class DialectOp(IRDLOperation):
+    """A dialect definition."""
+
+    name = "irdl.dialect"
+
+    sym_name: OpAttr[StringAttr]
+    body: SingleBlockRegion
+
+    def __init__(self, name: str | StringAttr, body: Region):
+        if isinstance(name, str):
+            name = StringAttr(name)
+        super().__init__(attributes={"sym_name": name}, regions=[body])
+
+    @classmethod
+    def parse(cls, parser: Parser) -> DialectOp:
+        sym_name = parser.parse_symbol_name()
+        region = parser.parse_optional_region()
+        if region is None:
+            region = Region(Block())
+        return DialectOp(sym_name, region)
+
+    def print(self, printer: Printer) -> None:
+        printer.print(" @", self.sym_name.data, " ")
+        if not self.body.block.is_empty:
+            printer.print_region(self.body)
+
+
+@irdl_op_definition
+class TypeOp(IRDLOperation):
+    """A type definition."""
+
+    name = "irdl.type"
+
+    sym_name: OpAttr[StringAttr]
+    body: SingleBlockRegion
+
+    traits = frozenset([HasParent(DialectOp)])
+
+    def __init__(self, name: str | StringAttr, body: Region):
+        if isinstance(name, str):
+            name = StringAttr(name)
+        super().__init__(attributes={"sym_name": name}, regions=[body])
+
+    @classmethod
+    def parse(cls, parser: Parser) -> TypeOp:
+        sym_name = parser.parse_symbol_name()
+        region = parser.parse_optional_region()
+        if region is None:
+            region = Region(Block())
+        return TypeOp(sym_name, region)
+
+    def print(self, printer: Printer) -> None:
+        printer.print(" @", self.sym_name.data, " ")
+        if not self.body.block.is_empty:
+            printer.print_region(self.body)
+
+
+@irdl_op_definition
+class AttributeOp(IRDLOperation):
```

Comment:
Interesting, the TypeOp and the AttributeOp look exactly the same. Funny how that goes...

Reply:
Actually, should we factorize them?

Reply:
It's as if they were the same thing, right?
I don't want to factorize them though, since we want to keep the exact same dialect as in MLIR.
I can create a `TypeOrAttributeOp` though.

Reply:
It's fine for me as-is, if this is the actual MLIR dialect. Just a general comment on the dialect...